### PR TITLE
fix #8457, #9606: introduce SAF Scoped Storage for public dirs/files into c:geo

### LINKED ISSUE #8457: Rework storage framework acc. to Android 11 requirements
- Install nightly build 2020.06.13-NB-9bbc28e
- Start app and apply permissions for storage and position
- Go to "About c:geo" and check system information tab

Result as expected:
 - `User storage c:geo dir: /storage/emulated/0/cgeo`

- Install nightly build 2020.06.14-NB-e549d24
- Start app and apply permissions for storage and position
- Go to "About c:geo" and check system information tab

Result (wrong):
- `User storage c:geo dir: /data/user/0/cgeo.geocaching`
This fallback should only occur in exceptional cases as this directory cannot be accessed by users (without root)

Remark:
The dir is not changed when upgrading from 2020.06.13 to 2020.06.14, the problem only occurs on installation but not on upgrade

This smells like being related to our change of `targetSDK` I would guess.

### LINKED ISSUE #9606: Migrate and use SAF for storing external user data (previous /cgeo/*)
Currently c:geo is storing external user data in `/cgeo/*` using the plain file access permission. 

Details about the current disk usage structure can be found here: https://github.com/cgeo/cgeo/wiki/Disk-usage-structure
We are talking about item <3> of the table in this wiki article (typically used for GPX imports and exports, logfiles, backups and in default case optionally also offline maps).  

Starting with target SDK=30 Android enforces to use the StorageAccessFramework (SAF) instead of plain file access permissions.
Therefore we should migrate from the current access mode into using the SAF.

The general approach has been discussed in #8457 already, which I try to summarize below (correct me, if I am wrong):
- Use SAF to request access to a base directory, which from then on is used by c:geo for storing the mentioned data and we simply use the structure below as we do today for `/cgeo/`.
- For fresh installations the user might set the base directory up to his own choice. This setup should be implemented into the installation wizard at some point (see #9589)
- In the upgrade case we recommend to the user to select the existing `/cgeo/` to be used as his base directory, which thereafter does not require any special migration. 

To be agreed/clarified (or I missed it in the original thread):
- Do we want to take care of any migration steps (i.e. copy/move files) if an existing users selects another base directory than `/cgeo/` or simply fire and forget the `/cgeo/` and let the user clean up?
- Currently we have a dir chooser to select GPX import and export directories. I would prefer to use this opportunity to get rid of these menus (and the ugly dir chooser) and simply focus everything into the selected base directory and the existing structure below (directory /gpx/). Users shall places their imports into this dir to use the internal GPX import function. Alternatively they can open GPX files with c:geo anyway. For exports we store them in our structure and offer sharing them (this is on the way or already implemented AFAIK).

### PR DESCRIPTION
**Pull request by eddiemuc**
Fix #9606 

It is time to end` storage_framework` and introduce scoped storage to `master`. I rebased it, squashed last commits and force-pushed to it (hopefully) one last time. 

Please as of now **do not use `storage_framework `any more**. For the remainder of the PR I would consider it my personal branch, and I will fix any complaints in it as if it were a branch in my personal fork. Hope that's ok. I think any other way will get very messy.

Concept is documented in wiki (https://github.com/cgeo/cgeo/wiki/Disk-Usage-migration-towards-Android-11) and review is handled by another issue (#9661)

As of now `storage_framework` contains complete SAF migration for c:geo with following exceptions:
* Offline log image: treated with separate issue #9609 
* Backups: treated with separate issue #9720 
* Spoilers: sep. issues #9719 
* Map Themes: trated with separate issue #9607 

Those exceptional cases will continue to work with the "old" file-style access after `storage_framework ` is merged.

Storage Framework can be tested before merge by installing the apk from branch `storage_framework`

_Originally posted by @eddiemuc in https://github.com/cgeo/cgeo/issues/9606#issuecomment-756048751_ (text above is modified)

### CODE REVIEW

**Review by fm-sys — COMMENTED** [review]
Thanks @eddiemuc for your great work! I admire very much how you got into the SAF topic. Unfortunately, I can't review your PR codewise but I added some comments...

Before merging we should IMHO do some basic tests with the SAF nightly and merge the PR afterwards. All other minor things are probably easier to fix at master branch. 

We should only think about if we want to start another small beta/feature release before merging...

File: main/src/cgeo/geocaching/CacheListActivity.java [hunk 1]
```diff
@@ -146,8 +145,8 @@
     private static final int MAX_LIST_ITEMS = 1000;
     private static final int REFRESH_WARNING_THRESHOLD = 100;
 
-    private static final int REQUEST_CODE_IMPORT_GPX = 1;
-    private static final int REQUEST_CODE_RESTART = 2;
+    //private static final int REQUEST_CODE_IMPORT_GPX = 1;
+    //private static final int REQUEST_CODE_RESTART = 2;
```
[the remark this answers is not in the document]
Comment (eddiemuc) [reply main/src/cgeo/geocaching/CacheListActivity.java:149]: yes, removed

File: main/src/cgeo/geocaching/utils/DebugUtils.java [hunk 2]
```diff
@@ -19,33 +21,31 @@
...
 
     private DebugUtils() {
         // utility class
     }
 
     public static void createMemoryDump(@NonNull final Context context) {
         Toast.makeText(context, R.string.init_please_wait, Toast.LENGTH_LONG).show();
-        final File file = FileUtils.getUniqueNamedLogfile("cgeo_dump", "hprof");
+        final File file = ContentStorage.get().createTempFile();
+        //final File file = FileUtils.getUniqueNamedLogfile("cgeo_dump", "hprof");
```
[the remark this answers is not in the document]
Comment (eddiemuc) [reply main/src/cgeo/geocaching/utils/DebugUtils.java:38]: yes, removed

File: main/src/cgeo/geocaching/settings/SettingsActivity.java [hunk 3]
```diff
@@ -178,15 +179,19 @@ private void initPreferences() {
         for (final int k : new int[] {
                 R.string.pref_pass_vote, R.string.pref_signature,
                 R.string.pref_mapsource, R.string.pref_renderthemepath,
-                R.string.pref_gpxExportDir, R.string.pref_gpxImportDir,
+                //R.string.pref_gpxExportDir, R.string.pref_gpxImportDir,
```
[the remark this answers is not in the document]
Comment (eddiemuc) [reply main/src/cgeo/geocaching/settings/SettingsActivity.java:182]: yes here too :-). removed

File: main/res/values/preference_keys.xml [hunk 4]
```diff
@@ -86,6 +86,7 @@
     <string translatable="false" name="pref_choose_list">choose_list</string>
     <string translatable="false" name="pref_mapsource">mapsource</string>
     <string translatable="false" name="pref_maplanguage">maplanguage</string>
+    <string translatable="false" name="pref_mapsforge_scale_text">mapsforgeScaleText</string>
```
[the remark this answers is not in the document]
Comment (eddiemuc) [reply main/res/values/preference_keys.xml:89]: yes, I think you are right. This must have been one of these merge sessions...
removed.

**Comment by eddiemuc** [comment]
Hi @fm-sys , many thanks for you review! You were right with all findings of course, and I fixed them all just now.

> Before merging we should IMHO do some basic tests with the SAF nightly and merge the PR afterwards. All other minor things are probably easier to fix at master branch.

Agreed. As far as I know @Lineflyer already set up a nightly for branch storage_framework some weeks ago which should be usable for that. Apart from that: this is a VERY big PR, so time may be needed for review too.

> We should only think about if we want to start another small beta/feature release before merging...

As far as I understand there is a release short before delivery. Merging of storage_framework to master should not occur before that imho.

**Comment by Lineflyer** [comment]
Nightlies for the SAF branch should be on: https://download.cgeo.org/cgeo-nightly-SAF.apk

IMHO we should wait for upcoming release, wait few days for possible urgent regressions, then merge master into release once again, then merge SAF.

Until then:
I will also start testing the nightly SAF including some upgrade scenarios next days and report back if I see any problems

File: main/src/cgeo/geocaching/storage/DocumentContentAccessor.java [hunk 5]
```diff
@@ -107,10 +107,10 @@ private Uri createInternal(@NonNull final Folder folder, @NonNull final String n
         final String docName = createUniqueFilename(name, queryDir(folderUri, new String[]{ DocumentsContract.Document.COLUMN_DISPLAY_NAME }, c -> c.getString(0)));
         try (ContextLogger cLog = new ContextLogger("DocumentFolderAccessor.create %s: %s", folder, name)) {
             //Do NOT pass a mimeType. It will then be selected based on the file suffix.
-            return DocumentsContract.createDocument(getContext().getContentResolver(), folderUri, null, docName);
+            //Use an empty string for mimeType; null triggers an exception in API21 and API30 (funnily enough: noz in API29)
```
Comment (bekuno) [inline main/src/cgeo/geocaching/storage/DocumentContentAccessor.java:110]: /noz/not/

Comment (eddiemuc) [reply main/src/cgeo/geocaching/storage/DocumentContentAccessor.java:110]: Damn! I will fix this with the next bug finding...

File: main/res/values/strings.xml [hunk 6]
```diff
@@ -840,6 +843,44 @@
...
+
+    <string name="persistablefolder_base">Base</string>
+    <string name="persistablefolder_offline_maps">Offline Maps</string>
+    <string name="persistablefolder_offline_maps_themes">Map Themes</string>
+    <string name="persistablefolder_gpx">GPX</string>
+    <string name="persistablefolder_backup">Backup</string>
+    <string name="persistablefolder_logfiles">Logfiles</string>
+    <string name="persistablefolder_fieldnotes">Field Notes</string>
+    <string name="persistablefolder_images">Log Images</string>
+
```
Comment (Lineflyer) [inline main/res/values/strings.xml:883]: Are these `persistablefolder_*` only used for the system information? In that case we don't need localization.

**Review by Lineflyer — COMMENTED** [review]
Looked briefly over it without really able to review the code.
I will keep testing the nightlies from `storage_framework` branch and report in dedicated issues.

File: main/res/values/strings.xml (hunk 6 above)
Comment (eddiemuc) [reply main/res/values/strings.xml:883]: Iirc also in the folder selection dialog title. But I have to check

**Comment by eddiemuc** [comment]
retest this please

**Comment by bekuno** [comment]
retest this please

File: main/res/values/strings.xml (hunk 6 above)
Comment (eddiemuc) [reply main/res/values/strings.xml:883]: Confirmed. It is used in the "select folder" dialogs like this:
image

Found no other places though.

Do you think that a non-translatable value is enough for this case?

**Comment by eddiemuc** [comment]
Question to all: due to changes in `master`, `storage_framework `has now again conflicts and needs a rebase. This will happen in the future again I assume.
May I continue to squash, rebase and force-push to `storage_framework `without disturbing anyone's work?

File: main/res/values/strings.xml (hunk 6 above)
Comment (Lineflyer) [reply main/res/values/strings.xml:883]: No, in that case I think it should be translated. So all fine here.

**Comment by Lineflyer** [comment]
To me it looks like it is working smooth despite the issue I already raised.
Tested on an Android 11 device as well as (roughly) on an Android 5.0.2 device. 

Still we should wait until the next branching from `master` to `release` has been done if this is a feasible thing given the need of frequent rebasing and conflict resolution.

**Comment by eddiemuc** [comment]
It makes sense to wait until release is stable and the current problems e.g. wrt vanishing caches are solved.

After that I don't see a reason not to merge SAF. It is a necessary thing to do before we can start migrating the missing things (themes, backup etc), also I am looking forward to get feedback from the real nightly. Also the change is now open for code review for a long time.

In case you want to wait longer, I kindly ask not to do any bigger feature changes in c:geo until this is merged, because the rebase-merges of the branch will kill me otherwise. Especially please no changes which somehow deal with files, settings or system information.

File: main/src/cgeo/geocaching/storage/DocumentContentAccessor.java (hunk 5 above)
Comment (eddiemuc) [reply main/src/cgeo/geocaching/storage/DocumentContentAccessor.java]: done

**Comment by eddiemuc** [comment]
Branch needs rebase. Since there are currently no open code review comments this feels like a good time to do this. I will thus now squash, rebase and force-push `storage_framework`. Please don't work against it meanwhile and refresh local branch afterwards

**Comment by eddiemuc** [comment]
Rebase & force-push is done

**Comment by eddiemuc** [comment]
Some time has passed now since this PR was created and some reviews and tests were done.

Depending on release and bug situation merging this PR to master might still need some time, but once this time has come I would consider this PR reviewed according to 4-eye-principle. Unless of course someone plans to do some more review work on this, in that case please give me a short notice.

**Comment by Lineflyer** [comment]
IMHO it has reached a very good and working level to continue on master branch.
I will take a look into current releases performance and regressions, then post a followup on the "next release" issue.

**Comment by Lineflyer** [comment]
`release` has been fast-forwarded and beta phase started.
So `master` would now be free to merge SAF IMHO.

**Comment by eddiemuc** [comment]
Ok so I would merge #9731 and #9734 in a few hours (so people have the chance to read this) if no one objects...

**Comment by moving-bits** [comment]
@eddiemuc 
Yes, go ahead, please. You should be the first to merge on the "new" `master` branch to avoid any further conflict resolution for those PR.

**Comment by eddiemuc** [comment]
ok, off we go...

**Comment by fm-sys** [comment]
Do we need to disable the SAF nightly built at Jenkins, as it is now merged to master?

**Comment by eddiemuc** [comment]
> Do we need to disable the SAF nightly built at Jenkins, as it is now merged to master? 

Good thinking, I guess we need to do that. But I don't know how

**Comment by Lineflyer** [comment]
> Do we need to disable the SAF nightly built at Jenkins, as it is now merged to master?

Thanks, disabled the job schedule and moved it into "z_Deprecated jobs" to be used as a template.

**Pull request merged**